keps/sig-network/2030-topology-aware-proxying/README.md

```diff
@@ -245,6 +253,138 @@ not reliant on any other feature being enabled in any other release. [KEP
+  transition process, just turning the feature gate on or off.
+
+### Rollout, Upgrade and Rollback Planning
+
+* **How can a rollout fail? Can it impact already running workloads?**
+  If the `for-zone` label is set incorrectly by KEP 2004 it is possible that
+  there may be no endpoints delivered to a zone for a Service. That would mean
+  that traffic originating from that zone would have nowhere to go. Importantly
+  this is not an issue if the label is not set, as all EndpointSlices without
+  these labels will continue to be consumed.
+
+* **What specific metrics should inform a rollback?**
+  Existing kube-proxy metrics can be used to determine the health of kube-proxy
+  and therefore this feature. Indicators include
+  `sync_proxy_rules_endpoint_changes_pending` consistently being too high or
+  `sync_proxy_rules_last_timestamp_seconds` being significantly older than the
+  latest changes to relevant EndpointSlices.
+
+* **Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?**
+  The feature has not been implemented yet but this will be tested manually.
+
+* **Is the rollout accompanied by any deprecations and/or removals of features,
+  APIs, fields of API types, flags, etc.?**
+  No.
+
+### Monitoring Requirements
+
+* **How can an operator determine if the feature is in use by workloads?**
+  If there are any EndpointSlices with the `for-zone` label set, KEP 2004 is
+  enabled. That is a prerequisite for this feature to have any impact. Enabling
+  this feature requires that the `TopologyAwareProxying` feature gate to be
+  enabled on kube-proxy.
+
+* **What are the SLIs (Service Level Indicators) an operator can use to
+  determine the health of the service?**
+  Any SLIs that already apply to kube-proxy should continue to apply here.
+
+* **What are the reasonable SLOs (Service Level Objectives) for the above SLIs?**
+  There's nothing particularly unique about this feature as far
+
+* **Are there any missing metrics that would be useful to have to improve
+  observability of this feature?**
+  Kube-Proxy already has a number of useful metrics, it's not clear that any
+  additional metrics would be useful here.
+
+### Dependencies
+
+* **Does this feature depend on any specific services running in the cluster?**
+  This feature has the same dependencies as kube-proxy itself, primarily the
+  EndpointSlice controller.
+
+
+### Scalability
+
+* **Will enabling / using this feature result in any new API calls?**
+  Depending on the results of testing, this may result in additional watch on
+  EndpointSlices. Fundamentally we want kube-proxy to watch all EndpointSlices
+  without a `for-zone` label or with one set to the zone of the node kube-proxy
+  is in. If separating this into 2 watches results in degraded performance or
+  too much complexity we will simply use the existing watch and filter the
+  EndpointSlices in kube-proxy.
```

Comment:
I'm almost sure this will cause problems if we don't add some customization in kube-apiserver.

Basically - note that without any changes, we will have to apply the selector for every watch now. Even though the selectors are pretty simple, if you multiple it by amount of watchers and events - this is non-negligible.

We can probably mitigate it by some simple indexes in the registry though - which might be the best option.

Also - why additional watch, not additional 2 watches?

IIUC, what we will be watching now would be:
- watch for "my zone"
- watch for "my region"
- watch for "objects without those labels"

So 3 instead of 1, right?

Reply:
Discussed below - a safer approach here seems to involve no changes to the watch and instead implementing filtering in kube-proxy. This allows us to fallback if no endpoints are delivered to a zone.

Reply:
My original thought was to start with zone only, but you're right that we'd eventually end up with 3 watches. I've updated the approach to use existing watch and fallback when endpoints aren't present.

Reply:
Please update this point to reflect this too :)

Reply:
Updated, thanks!